File: Leetcode/Array/11. Container With Most Water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int ans = solve(height);

        reverse(height.begin(), height.end());

        ans = max(ans, solve(height));

        return ans;
    }

    int solve(vector <int> &v) {
        int n = v.size(), res = 0;
        vector <int> mx(n);

        for(int i = n-1; i >= 0; i--) mx[i] = (i == n-1) ? v[i] : max(mx[i+1], v[i]);

        for(int i = 0; i < n; i++) {
            int st = i+1, en = n-1, area = 0;

            while(st <= en) {
                int mid = (st + en) / 2;

                if(mx[mid] >= v[i]) {
                    area = v[i] * (mid - i);
                    st = mid + 1;
                }
                else en = mid - 1;
            }

            res = max(res, area);
        }

        return res;
    }
};

File: Leetcode/Practice/Binary Search/11. Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        ans, n = 0, len(height)
        mx = [[0 for _ in range(2)] for _ in range(n)]

        for i in range(n):
            mx[i][0] = height[i] if i == 0 else max(height[i], mx[i-1][0])
        
        for i in range(n-1, -1, -1):
            mx[i][1] = height[i] if i == n-1 else max(height[i], mx[i+1][1])

        for i in range(n):
            st, en, pos = 0, i, i
            while st <= en:
                mid = (st + en) // 2
                if mx[mid][0] >= height[i]:
                    pos = mid
                    en = mid - 1
                else:
                    st = mid + 1

            area = (i - pos) * height[i]
            ans = max(ans, area)

            st, en, pos = i, n-1, i
            while st <= en:
                mid = (st + en) // 2
                if mx[mid][1] >= height[i]:
                    pos = mid
                    st = mid + 1
                else:
                    en = mid - 1
            
            area = (pos - i) * height[i]
            ans = max(ans, area)
        
        return ans